### DISCUSSION TITLE
Differentiability of Descriptors

### DISCUSSION THREAD

**Discussion by completementgaga** [kornia/kornia#3175] [Potential Issue]
Good morning everybody,

I am trying to use Kornia Features to set a 2D/3D registration scheme. Namely I want to align a 2D image to the most similar planar slice of the 3D image. The 2D image might be slightly deformed (in a non homographic way, consider rather a free form deformation) compared to its best matching slice from he 3D image, so I am interested in local descriptors, robust to conformal maps, like SIFT.

However, I face difficulties to guess what is the most relevant model in Kornia.features for this task. Particularly, I get some weird results/errors when I try to differentiate losses defined in terms of the descriptors given by 
 - kornia.feature.DeDoDe
 - kornia.feature.SIFTFeatureScaleSpace
 
Please, can anyone tell me what part of these pipelines are differentiable ? My hope is that, once the key points are fixed, the description functions are differentiable. For DeDoDe, I would say from the paper that the whole pipeline should be differentiable.

I must say I had some success in differentiating with the DISK module, but it is not clear to me if it robust enough to image deformations. Any hint about that ? What do you think of the relevance of RoMa for the kind of matching I want to perform?

Any help on any of the above points would be greatly appreciated,

Best,
Gaël.

**Comment by RizgarOzan** [comment]
I checked this on current main (`320f4611`, torch 2.14, CPU), for anyone who lands here.

**DeDoDe:** `DeDoDe.detect()` and `DeDoDe.describe()` are decorated with `@torch.inference_mode()` (`kornia/feature/dedode/dedode.py`), and `forward()` calls both. So everything they return is an inference tensor with no graph, and `loss.backward()` fails with `element 0 of tensors does not require grad and does not have a grad_fn`. That is the error you most likely hit. It is not about the method itself: if you call the undecorated function, the descriptors are differentiable with respect to the image:

```python
kps = ...  # (B, N, 2) in [-1, 1], e.g. from dedode.detect(), or fixed
desc = type(dedode).describe.__wrapped__(dedode, images, kps)
desc.sum().backward()  # works, images.grad is finite
```

I checked this with the `B` descriptor: same values as `describe()`, finite non-zero gradient. With the `G` descriptor, the DINOv2 part runs inside its own `torch.inference_mode()` block (`kornia/feature/dedode/encoder.py`) and is cloned afterwards, so the gradient only flows through the VGG branch, not through DINOv2. Keypoint *detection* is sampling from the score map, so the keypoint positions themselves carry no gradient either way.

**SIFTFeatureScaleSpace:** the descriptors do carry a gradient to the image. On a random 96x96 image, `lafs, resp, descs = KF.SIFTFeatureScaleSpace(num_features=50)(img)` followed by `descs.sum().backward()` gives a finite gradient. With the keypoints fixed, `KF.get_laf_descriptors(img, lafs.detach(), KF.SIFTDescriptor(32, rootsift=True))` also backpropagates, and `torch.autograd.gradcheck` passes for `SIFTDescriptor` on float64 patches. If you get odd values there, a small sample would help narrow it down.

I have not compared the robustness of DISK, DeDoDe or RoMa to non-rigid deformation, so I can't say which one fits your case best.

(AI note: I used Claude to write and run the checks above; the claims are from running them on main and reading the code linked.)
